pubsub/rabbitmq/rabbitmq.go

```diff
@@ -44,89 +92,199 @@ func (r *rabbitMQ) Init(metadata pubsub.Metadata) error {
+	if channel != nil {
+		return channel, connectionCount, nil
+	}
+
+	r.logger.Warnf("%s reconnecting ...", logMessagePrefix)
+	err := r.reconnect(connectionCount)
+	if err != nil {
+		return nil, connectionCount, err
+	}
+
+	channel, connectionCount = r.getChannel()
+
+	return channel, connectionCount, nil
+}
+
 func (r *rabbitMQ) Publish(req *pubsub.PublishRequest) error {
-	err := r.ensureExchangeDeclared(req.Topic)
+	channel, connectionCount, err := r.getChannelOrReconnect()
+	if err != nil {
+		return err
+	}
+
+	err = r.ensureExchangeDeclared(channel, req.Topic)
 	if err != nil {
 		return err
 	}
 
 	r.logger.Debugf("%s publishing message to topic '%s'", logMessagePrefix, req.Topic)
 
-	err = r.channel.Publish(req.Topic, "", false, false, amqp.Publishing{
+	err = channel.Publish(req.Topic, "", false, false, amqp.Publishing{
 		ContentType:  "text/plain",
 		Body:         req.Data,
 		DeliveryMode: r.metadata.deliveryMode,
 	})
 
 	if err != nil {
+		if mustReconnect(channel, err) {
+			r.logger.Warnf("%s pubsub publisher for %s is reconnecting ...", logMessagePrefix, req.Topic)
+			r.reconnect(connectionCount)
+		}
+
 		return err
 	}
 
 	return nil
 }
 
 func (r *rabbitMQ) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
-	err := r.ensureExchangeDeclared(req.Topic)
+	queueName := fmt.Sprintf("%s-%s", r.metadata.consumerID, req.Topic)
+
+	go r.subscribeForever(req, queueName, handler)
+
+	return nil
+}
+
+func (r *rabbitMQ) prepareSubscription(channel rabbitMQChannelBroker, req pubsub.SubscribeRequest, queueName string) (*amqp.Queue, error) {
+	err := r.ensureExchangeDeclared(channel, req.Topic)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	queueName := fmt.Sprintf("%s-%s", r.metadata.consumerID, req.Topic)
-
 	r.logger.Debugf("%s declaring queue '%s'", logMessagePrefix, queueName)
-	q, err := r.channel.QueueDeclare(queueName, true, r.metadata.deleteWhenUnused, false, false, nil)
+	q, err := channel.QueueDeclare(queueName, true, r.metadata.deleteWhenUnused, false, false, nil)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	r.logger.Debugf("%s binding queue '%s' to exchange '%s'", logMessagePrefix, q.Name, req.Topic)
-	err = r.channel.QueueBind(q.Name, "", req.Topic, false, nil)
+	err = channel.QueueBind(q.Name, "", req.Topic, false, nil)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	msgs, err := r.channel.Consume(
-		q.Name,
-		queueName,          // consumerId
-		r.metadata.autoAck, // autoAck
-		false,
-		false, // noLocal
-		false, // noWait
-		nil,
-	)
-	if err != nil {
-		return err
-	}
+	return &q, nil
+}
 
-	go r.listenMessages(msgs, req.Topic, handler)
+func (r *rabbitMQ) subscribeForever(
+	req pubsub.SubscribeRequest,
+	queueName string,
+	handler func(msg *pubsub.NewMessage) error) {
+	var err error
+	var connectionCount int
+	var channel rabbitMQChannelBroker
+	var q *amqp.Queue
+
+	for {
+		err = nil
+		for {
```

Comment:
don't we want to sleep here? also, does this mean we consistently subscribe to a queue name? if so, won't it cause messages to be delivered multiple times?

Reply:
I did not see messages being delivered multiple times, I will check that. Line 260 is blocking and will keep consuming messages until an error happens.

Reply:
The only reason this inner `for` loop will cycle is if for some reason `r.listenMessages` (line 260) exists without an error code - which could be an odd behavior.

Reply:
Here is the scenario:

Publisher publishes messages in sequence - if disconnected, publisher will retry the same message over and over.
Subscriber, prints each message.
socat as TCP proxy between dapr sidecar and rabbitMQ (so simulate disconnection without service restart)


app ----> daprd -----> socat ----> rabbitmq

I let the publisher and subscriber run for some time and then I stop socat process to simulate disconnection. Then I wait a few seconds and bring socat proxy back. Publisher recovers within 1 second (sleep time between retries in the app) while Subscriber takes about 10 seconds to recover. As a result, the subscriber comes back and misses some of the messages sent by the publisher.

```txt
== APP == Subscriber got message: This is message #13

== APP == 2020-11-06 14:50:40,608 {HH:mm:ss.SSS} [http-nio-3000-exec-7] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@4070db10, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604703040598569652, hasEnded=true}

== APP == Subscriber got message: This is message #14

== APP == 2020-11-06 14:50:41,240 {HH:mm:ss.SSS} [http-nio-3000-exec-8] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@4f293ff4, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604703041221124104, hasEnded=true}

== APP == Subscriber got message: This is message #15

== APP == 2020-11-06 14:50:41,878 {HH:mm:ss.SSS} [http-nio-3000-exec-9] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@397e4143, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604703041867191892, hasEnded=true}

== DAPR == time="2020-11-06T14:50:42.4268-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, Exception (504) Reason: \"channel/connection is not open\"" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:50:52.428049-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:50:52.428101-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, channel not initialized" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:51:02.433258-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:51:02.434015-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, channel not initialized" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:51:12.434737-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:51:12.435603-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, channel not initialized" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:51:22.436236-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-06T14:51:22.463798-08:00" level=info msg="rabbitmq pub/sub: connected" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== APP == Subscriber got message: This is message #33

== APP == 2020-11-06 14:51:22,952 {HH:mm:ss.SSS} [http-nio-3000-exec-2] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@6faab040, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604703082937947798, hasEnded=true}

```

Reply:
This (BTW) is the same behavior if the Subscriber's daprd is restarted but the publisher keeps publishing. So, this change is not adding a new side effect of a restart.

Reply:
I found the issue. It is because Auto-delete (deletedWhenUnused) is enabled by default in this implementation, so the queue is recreated when the only consumer reconnects. This is fixed with a config change in the test. Reference: https://www.rabbitmq.com/queues.html#properties

New output:

```txt
== APP == Subscriber got message: This is message #8

== APP == 2020-11-09 12:20:52,525 {HH:mm:ss.SSS} [http-nio-3000-exec-2] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@7f8652e2, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604953252520346049, hasEnded=true}

== APP == Subscriber got message: This is message #9

== APP == 2020-11-09 12:20:52,583 {HH:mm:ss.SSS} [http-nio-3000-exec-1] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@4415ba50, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604953252578431937, hasEnded=true}

== DAPR == time="2020-11-09T12:20:53.126738-08:00" level=warning msg="rabbitmq pub/sub: declaring queue 'subscriber-testingtopic' autodelete: false" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:20:53.126831-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, Exception (504) Reason: \"channel/connection is not open\"" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:03.130707-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:03.130784-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, channel not initialized" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:13.134888-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:13.135684-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, channel not initialized" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:23.136081-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:23.136736-08:00" level=error msg="rabbitmq pub/sub: error in subscription for subscriber-testingtopic, channel not initialized" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:33.137124-08:00" level=warning msg="rabbitmq pub/sub: pubsub subscription for subscriber-testingtopic is reconnecting ..." app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:33.14933-08:00" level=info msg="rabbitmq pub/sub: connected" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== DAPR == time="2020-11-09T12:21:33.150707-08:00" level=warning msg="rabbitmq pub/sub: declaring queue 'subscriber-testingtopic' autodelete: false" app_id=subscriber instance=MININT-UEMT0H6.northamerica.corp.microsoft.com scope=dapr.contrib type=log ver=edge

== APP == Subscriber got message: This is message #10

== APP == 2020-11-09 12:21:33,163 {HH:mm:ss.SSS} [http-nio-3000-exec-4] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@5eef5416, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604953293158189854, hasEnded=true}

== APP == Subscriber got message: This is message #11

== APP == 2020-11-09 12:21:33,170 {HH:mm:ss.SSS} [http-nio-3000-exec-5] INFO  i.o.e.logging.LoggingSpanExporter - span: SpanWrapper{delegate=io.opentelemetry.sdk.trace.RecordEventsReadableSpan@752b1858, resolvedLinks=[], resolvedEvents=[], attributes={handler=AttributeValueString{stringValue=afterCompletion}}, totalAttributeCount=2, totalRecordedEvents=0, status=Status{canonicalCode=OK, description=null}, name=/testingtopic, endEpochNanos=1604953293166395490, hasEnded=true}

```


Config:
```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: messagebus
spec:
  type: pubsub.rabbitmq
  metadata:
    - name: host
      value: amqp://localhost:5670
    - name: consumerID
      value: myConsumerID
    - name: autoAck
      value: "false"
    - name: deletedWhenUnused
      value: "false"
```